### PR TITLE
Deprecate client's abstract methods

### PR DESCRIPTION
**Pull request by jderusse**
Following #605

### CODE REVIEW

File: src/Core/src/Sts/StsClient.php [hunk 1]
```diff
@@ -213,16 +193,22 @@ protected function getEndpointMetadata(?string $region): array
 
     protected function getServiceCode(): string
     {
+        @trigger_error('Using the client with an old version of Core is deprecated. Run "composer update async-aws/core".', \E_USER_DEPRECATED);
```
Comment (jderusse) [inline src/Core/src/Sts/StsClient.php:196]: Shouldn't we use `core:^1.2` instead and removes those methods?

File: src/Service/CloudFormation/src/CloudFormationClient.php [hunk 2]
```diff
@@ -176,16 +158,22 @@ protected function getEndpointMetadata(?string $region): array
 
     protected function getServiceCode(): string
```
Comment (Nyholm) [inline src/Service/CloudFormation/src/CloudFormationClient.php:159]: This function can be removed in all clients but the one tagged 1.0

File: src/Core/src/AbstractApi.php [hunk 3]
```diff
@@ -83,11 +84,29 @@ public function __construct($configuration = [], ?CredentialProvider $credential
...
+    protected function getServiceCode(): string
+    {
+        throw new LogicException(\sprintf('The method "%s" should not be called. The Client "%s" must implement the "%s" method.', __FUNCTION__, \get_class($this), 'getEndpointMetadata'));
+    }
 
-    abstract protected function getSignatureVersion(): string;
+    /**
+     * @deprecated
+     */
+    protected function getSignatureVersion(): string
```
Comment (Nyholm) [inline src/Core/src/AbstractApi.php:98]: We can just remove these and update the version constraints. Right?

Comment (jderusse) [reply src/Core/src/AbstractApi.php:98]: yes, in all but 1.0 clients (because they may have been extends right?)

**Comment by Nyholm** [comment]
So after thinking of this for a while:

We cannot remove protected methods. In the AbstractApi we need to implement them and trigger deprecation. 

I’m all clients but SQS, s3 and ses, we can remove the override of those functions and require core 1.2. 

In SQS, s3 and ses, we must keep the override of these functions. 

I’m not sure if the abstractApi should throw exception or just trigger deprecation... hm. I think throwing is okey because all children has implemented these functions (they are currently abstract).

File: src/CodeGenerator/src/Generator/ClientGenerator.php [hunk 4]
```diff
@@ -67,12 +61,17 @@ public function getVersion() {
...
+            return strtr('        return [
+                "endpoint" => "ENDPOINT",
+                "signRegion" => REGION,
+                "signService" => SIGN_SERVICE,
+                "signVersions" => SIGN_VERSIONS,
+            ];' . "\n", [
+                'ENDPOINT' => strtr($config['endpoint'], ['%region%' => '$region']),
+                'REGION' => isset($config['signRegion']) ? \var_export($config['signRegion'], true) : '$region',
+                'SIGN_SERVICE' => var_export($config['signService'], true),
+                'SIGN_VERSIONS' => \json_encode($signatureVersions),
```
Comment (Nyholm) [inline src/CodeGenerator/src/Generator/ClientGenerator.php:73]: Haha. This is an ugly trick. I thought about it on #605, but I wasn’t sure it was pretty enough. :)

I’m okey with it :)

File: src/Service/CloudFormation/src/CloudFormationClient.php (hunk 2 above)
Comment (jderusse) [reply src/Service/CloudFormation/src/CloudFormationClient.php]: thanks, done

File: src/CodeGenerator/src/Generator/ClientGenerator.php (hunk 4 above)
Comment (jderusse) [reply src/CodeGenerator/src/Generator/ClientGenerator.php:73]: hey hey

**Comment by jderusse** [comment]
We ended to the same conclusion @Nyholm .
- remove method for clients <1
- deprecate methods for client >= 1
- convert abstract method to method that report the change.

I choose to throw an exception instead of triggering error because nobody have a reason to call those method:
- people that implement a client with core < 1.2 had overridden the method
- people that implement a client with core >= 1.2 should implement the new getEndpointMetadata and don't have to use those methods

**Pull request merged**